beps/bep_0011.rst

```diff
@@ -87,9 +87,37 @@ An added contact must not be dropped in the same message.
 
 Except for the initial PEX message the combined amount of added v4/v6 contacts should not exceed 50 entries. The same applies to dropped entries.
 
+    **Implementation Note**: If these limits conflict with large connection pools then implementations can calculate the canonical peer priorities from the perspective of a remote peer to choose which subset of
+connections to prioritize for inclusion in a message.  
+
 Messages must contain at least one of the following fields: ``added, added6, dropped, dropped6``.
 
-Clients may disconnect peers that egregiously violate these constraints.  
+Clients may disconnect peers that egregiously violate these constraints.
+
+
+Filling underpopulated lists
+============================
+
+A combination of the 
+
+* above liveness requirements
+* disconnecting seeds or partial seeds [#BEP-21]_ while seeding
+* disconnecting duplicate peers based on peer IDs between ipv4 and ipv6
+
+can lead to underpopulated ``added`` or ``added6`` lists. This problem frequently arises in swarms dominated by seeds where the seeds have no live connections to propagate and the peers
+only have seeds they can propagate to other seeds, dramatically reducing PEX's effectiveness. Similarly this can make it difficult to obtain ipv6 peers in an ipv4 dominated swarm because ipv6
+connections will be considered duplicates of already-established v4 ones, thus preventing their propagation via PEX.
+
+To remedy these issues the liveness requirement is relaxed if a client is connected to less than 25 clients for a particular address family. In that situation it may keep a list
+of most recently established connections for that address family and record the disconnect reason for those connections. If a remote client has been disconnected due to
+peer ID deduplication and the duplicate was on a different address family or because the remote was a seed then up to 25 of those may be included in the initial ``added`` or ``added6`` lists.
```

Comment:
I wonder if it makes sense to be a bit more vague about "because the remote was a seed". In practice it boils down to two peers that are not interested in any pieces the other peer has. Normally this happens because both peers are seeds, but it could also happen because both are partial seeds, or because one is a partial seed and the downloading peer also has all those pieces. I wonder if there may be other cases of disconnecting good peers that perhaps also should be added to a "recently connected" list of good peers.

Reply:
We can do that and make a non-exhaustive list of examples what is considered a disconnected good peer instead. Too vague without such a list might miss the goal here.

And I already referenced partial seeding case.

Reply:
I did a quick scan over libtorrent's disconnect reasons and I didn't see any obvious candidates other than what's already been mentioned. In any case there are clearly only a few disconnect conditions which account for the vast majority of eligible peers, so listing them is fine.